license-maven-plugin-git/src/main/java/com/mycila/maven/plugin/license/git/CopyrightAuthorProvider.java

```diff
@@ -56,12 +56,15 @@ public CopyrightAuthorProvider() {
     try {
       Map<String, String> result = new HashMap<String, String>(3);
       GitLookup gitLookup = getGitLookup(document.getFile(), properties);
-
+      
+      if (mojo.failIfShallow && gitLookup.isShallowRepository()) {
+        throw new RuntimeException("Shallow repository detected.");
+      }
```

Comment:
After some thinking, I am wondering if this is the right way of implementing that...

We should keep backward compatibility. So for users having a “working” or “acceptable” setup with shallow clones, it should continue to work as-is. The current situation can be accepted in many cases where the "badly" computed values would turn out matching the ones of the first commit.

For the remaining **new** users, I would probably avoid throwing and instead returning a Map, either empty or without the properties that cannot be correctly computed, and if a property cannot be computed, then log a message as info, if the user wants.

This would be a combination of 2 settings (we can find better names of course):

- `mojo.shallowSupport` => default to false, Activate or not correct support for shallow clones
- `mojo.warnUnresolvableProperties` => default to false, If `mojo.shallowSupport` is set to true and this property set to true, then we would log a warning for all the properties that are skipped because they cannot be computed
- `mojo.failOnUnresolvableProperties` => default to false, If `mojo.shallowSupport` is set to true and this property set to true, then we would fail the build if a property cannot be computed because of a shallow clone

I'm not sure a lot of people would activate 3. Only 1 and 2 would be enough. If someone is using a property that cannot be computed, it will be seen as the license check will fail the build.

Reply:
> We should keep backward compatibility. So for users having a “working” or “acceptable” setup with shallow clones, it should continue to work as-is.

It still will, if that property is set false; users can quickly "disable" this if desired.  This does go back to @hazendaz comment [here](https://github.com/mathieucarbou/license-maven-plugin/issues/253#issuecomment-1013783799_) about flipped defaults, although the context of that was a warning rather than an exception.

I do like the gist of your suggestion, offering 3 possible outcomes based on setting:
 - Completely ignore with no warning
 - Warn only
 - Fail build

Personally I'd like to add a fourth option which I think would be ideal if we can figure out how:
 - Warn about possible date inaccuracy only on failed header check (default true); but if license is not missing/rewritten, remain silent

The reason I like this:
 - It fits my (and probably a very common) use case: develop on a machine with full git history, but execute tests on a machine with shallow history.  I include the `format` goal in my local builds so all the headers are supposed to be rewritten when I commit; there's no reason CI should bother rewriting them again.
 - it assists in debugging current behavior. In my initial testing, I ended up with tons of header rewrites only on one machine, and having a log warning "I'm rewriting these because I detected a changed date but the repo is shallow" would have saved me some debugging time.   

With my current solution, I could quickly replace the failure with a warning but it would warn on all files, probably too much logspam and would be "off" by default.  But a warning associated with every header rewrite I think is reasonable to default "on".

This leads to your "map" suggestion:
> a Map, either empty or without the properties that cannot be correctly computed

Except I think the best purpose/name of the map would not say "cannot be computed" but "not certain it was computed correctly".  This would fit my use case above (develop full depth, CI test shallow) where it is likely that the headers are already correct, so the ideal outcome in that case would be:
 - If an "uncertain" property exists and the header is present, assume it's accurate, don't try to change it if the repo is shallow, and silently ignore
 - If an "uncertain" property exists and the header is missing, add it, and warn about possible year in accuracy (by default)

Or, one other solution, going back to my initial submission, "warnOnShallow".  I think it's **entirely reasonable to have a (default true) single log warning that a repository is shallow**.  It can be generated in the `GetLookup` instantiation so it only occurs once (basically warning instead of generating the `shallow` field in the current iteration), and only if the git properties are used.  
 - This still helps debugging (in the case of mass header check warnings/rewrites) but doesn't otherwise change the existing behavior.
 - Users have existing workarounds (adding `-d"license.skip"` or using profiles in CI) to respond to this warning; and/or toggle it off.

> This general behavior then could be applied to the SVN plugin

And I'll figure out how to detect a "sparse" SVN history and do something similar there.

Reply:
Whatever the solution : it **has** to be backward compatible and not fail existing setups or even log a lot of warnings in case of a project having a lot of modules.

The idea of having multiple properties matches what's done in other plugins too: warnIf..., failIf...

Reply:
> Whatever the solution : it **has** to be backward compatible and not fail existing setups or even log a lot of warnings in case of a project having a lot of modules.

OK, I think I'm pretty happy with the current iteration of it.  
1. It's backwards compatible, not failing existing setups
2. It doesn't log a lot of warnings. It only logs _one_.
  - And it only logs in the case that it's relevant, if the author info or creation/last modify year is requested.
  - It's therefore informational and useful for debugging 
3. Users can turn off that (one) warning line if they choose and "know what they're doing" with limited history.